File: results/2026-04-02_163146/tasks/11-semantic-version-bumper/csharp-script-sonnet/generated-code/bump-version.cs
// bump-version.cs — Semantic Version Bumper
// .NET 10 file-based app: run with `dotnet run bump-version.cs`
//
// Usage:
//   dotnet run bump-version.cs <version-file> <commits-file>
//   dotnet run bump-version.cs version.txt commits.json
//   dotnet run bump-version.cs package.json commits.json
//
// Where:
//   <version-file>  — path to a plain version.txt or package.json
//   <commits-file>  — path to a JSON array of conventional commit strings
//
// Output (stdout):
//   The new version string (e.g. "2.0.0")
//   A markdown changelog entry to stderr is also written to a CHANGELOG.md snippet
//
// Exit codes:  0 = success,  1 = error

using System.Text.Json;
using System.Text.RegularExpressions;

// ─────────────────────────────────────────────────────────────────────────────
// Entry point — top-level statements
// ─────────────────────────────────────────────────────────────────────────────

if (args.Length < 2)
{
    Console.Error.WriteLine("Usage: dotnet run bump-version.cs <version-file> <commits-file>");
    Console.Error.WriteLine();
    Console.Error.WriteLine("  <version-file>  — version.txt or package.json");
    Console.Error.WriteLine("  <commits-file>  — JSON array of conventional commit messages");
    Console.Error.WriteLine();
    Console.Error.WriteLine("Example:");
    Console.Error.WriteLine("  dotnet run bump-version.cs version.txt commits.json");
    Environment.Exit(1);
}

var versionFilePath = args[0];
var commitsFilePath = args[1];

try
{
    // 1. Read and parse current version
    if (!File.Exists(versionFilePath))
    {
        Console.Error.WriteLine($"Error: version file not found: '{versionFilePath}'");
        Environment.Exit(1);
    }

    var versionContent = await File.ReadAllTextAsync(versionFilePath);
    var currentVersion = AppVersionParser.Parse(versionContent, Path.GetFileName(versionFilePath));

    // 2. Load commit messages
    if (!File.Exists(commitsFilePath))
    {
        Console.Error.WriteLine($"Error: commits file not found: '{commitsFilePath}'");
        Environment.Exit(1);
    }

    var commitsJson    = await File.ReadAllTextAsync(commitsFilePath);
    var commitMessages = JsonSerializer.Deserialize<string[]>(commitsJson);

    if (commitMessages is null || commitMessages.Length == 0)
    {
        Console.Error.WriteLine("Warning: no commits found in commits file. Nothing to bump.");
        Console.WriteLine(currentVersion);
        Environment.Exit(0);
    }

    // 3. Analyse commits — determine bump type
    var parsedCommits = AppCommitAnalyzer.ParseCommits(commitMessages);
    var bumpType      = AppCommitAnalyzer.AnalyzeCommits(commitMessages);

    if (bumpType == AppBumpType.None)
    {
        Console.Error.WriteLine($"No releasable commits found. Current version: {currentVersion}");
        Console.WriteLine(currentVersion);
        Environment.Exit(0);
    }

    // 4. Calculate new version
    var nextVersion = AppVersionBumper.Bump(currentVersion, bumpType);

    // 5. Update the version file
    var updatedContent = AppVersionParser.UpdateContent(versionContent, Path.GetFileName(versionFilePath), nextVersion);
    await File.WriteAllTextAsync(versionFilePath, updatedContent);

    // 6. Generate changelog entry
    var changelog = AppChangelogGenerator.Generate(nextVersion, parsedCommits, DateTime.UtcNow);

    // 7. Append to CHANGELOG.md (create if it doesn't exist)
    var changelogPath = Path.Combine(Path.GetDirectoryName(versionFilePath) ?? ".", "CHANGELOG.md");
    if (File.Exists(changelogPath))
    {
        var existing = await File.ReadAllTextAsync(changelogPath);
        // Insert new entry after the first line (title) if it has one, otherwise prepend
        var lines = existing.Split('\n');
        if (lines.Length > 1 && lines[0].StartsWith("# "))
        {
            await File.WriteAllTextAsync(changelogPath, lines[0] + "\n\n" + changelog + "\n" + string.Join("\n", lines.Skip(1)).TrimStart());
        }
        else
        {
            await File.WriteAllTextAsync(changelogPath, changelog + "\n" + existing);
        }
    }
    else
    {
        await File.WriteAllTextAsync(changelogPath, "# Changelog\n\n" + changelog);
    }

    Console.Error.WriteLine($"Version bumped: {currentVersion} → {nextVersion} ({bumpType})");
    Console.Error.WriteLine($"Changelog written to: {changelogPath}");

    // Output the new version to stdout for scripting
    Console.WriteLine(nextVersion);
}
catch (Exception ex)
{
    Console.Error.WriteLine($"Error: {ex.Message}");
    Environment.Exit(1);
}

// ─────────────────────────────────────────────────────────────────────────────
// Domain models (self-contained copy for the file-based app)
// ─────────────────────────────────────────────────────────────────────────────

/// <summary>Immutable semantic version (MAJOR.MINOR.PATCH).</summary>
public record AppSemanticVersion(int Major, int Minor, int Patch)
{
    public override string ToString() => $"{Major}.{Minor}.{Patch}";

    public static AppSemanticVersion Parse(string version)
    {
        var trimmed = version.Trim();
        var parts   = trimmed.Split('.');

        if (parts.Length != 3)
            throw new FormatException(
                $"Invalid semantic version: '{trimmed}'. Expected MAJOR.MINOR.PATCH.");

        if (!int.TryParse(parts[0], out var major) ||
            !int.TryParse(parts[1], out var minor) ||
            !int.TryParse(parts[2], out var patch))
            throw new FormatException(
                $"Invalid semantic version: '{trimmed}'. All parts must be integers.");

        return new AppSemanticVersion(major, minor, patch);
    }
}

/// <summary>Version bump magnitude.</summary>
public enum AppBumpType { None = 0, Patch = 1, Minor = 2, Major = 3 }

/// <summary>A parsed conventional commit.</summary>
public record AppCommit(
    string Type,
    string Description,
    bool IsBreaking,
    string? Scope = null,
    string? Body  = null);

// ─────────────────────────────────────────────────────────────────────────────
// VersionParser
// ─────────────────────────────────────────────────────────────────────────────

public static class AppVersionParser
{
    private static readonly Regex PkgVersionRegex = new(
        @"""version""\s*:\s*""(\d+\.\d+\.\d+)""",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    public static AppSemanticVersion Parse(string content, string fileName) =>
        IsPackageJson(fileName)
            ? ParsePackageJson(content)
            : AppSemanticVersion.Parse(content.Trim());

    public static string UpdateContent(string content, string fileName, AppSemanticVersion v) =>
        IsPackageJson(fileName)
            ? UpdatePackageJson(content, v)
            : v.ToString();

    private static bool IsPackageJson(string name) =>
        name.Equals("package.json", StringComparison.OrdinalIgnoreCase);

    private static AppSemanticVersion ParsePackageJson(string content)
    {
        using var doc = JsonDocument.Parse(content);
        if (!doc.RootElement.TryGetProperty("version", out var el))
            throw new InvalidOperationException("package.json has no 'version' field.");
        return AppSemanticVersion.Parse(el.GetString()
            ?? throw new InvalidOperationException("'version' in package.json is null."));
    }

    private static string UpdatePackageJson(string content, AppSemanticVersion v)
    {
        if (!PkgVersionRegex.IsMatch(content))
            throw new InvalidOperationException("Could not find 'version' field in package.json.");
        return PkgVersionRegex.Replace(content, $@"""version"": ""{v}""");
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// CommitAnalyzer
// ─────────────────────────────────────────────────────────────────────────────

public static class AppCommitAnalyzer
{
    private static readonly Regex ConventionalRx = new(
        @"^(?<type>[a-zA-Z]+)(?:\((?<scope>[^)]+)\))?(?<bang>!)?\s*:\s*(?<desc>.+)$",
        RegexOptions.Compiled);

    private static readonly Regex BreakingBodyRx = new(
        @"BREAKING[\s\-]CHANGE\s*:",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    public static AppCommit ParseCommit(string message)
    {
        if (string.IsNullOrWhiteSpace(message))
            return new AppCommit("other", message ?? "", false, Body: message);

        var firstLine = message.Split('\n')[0].Trim();
        var m = ConventionalRx.Match(firstLine);
        if (!m.Success) return new AppCommit("other", firstLine, false, Body: message);

        var type      = m.Groups["type"].Value.ToLowerInvariant();
        var scope     = m.Groups["scope"].Success ? m.Groups["scope"].Value : null;
        var desc      = m.Groups["desc"].Value.Trim();
        var isBreak   = m.Groups["bang"].Success || BreakingBodyRx.IsMatch(message);

        return new AppCommit(type, desc, isBreak, scope, message);
    }

    public static IReadOnlyList<AppCommit> ParseCommits(IEnumerable<string> messages)
        => messages.Select(ParseCommit).ToList().AsReadOnly();

    public static AppBumpType AnalyzeCommits(IEnumerable<string> messages)
    {
        var result = AppBumpType.None;
        foreach (var msg in messages)
        {
            var bump = GetBump(ParseCommit(msg));
            if (bump > result) result = bump;
            if (result == AppBumpType.Major) break;
        }
        return result;
    }

    private static AppBumpType GetBump(AppCommit c) =>
        c.IsBreaking ? AppBumpType.Major :
        c.Type == "feat" ? AppBumpType.Minor :
        c.Type is "fix" or "perf" ? AppBumpType.Patch :
        AppBumpType.None;
}

// ─────────────────────────────────────────────────────────────────────────────
// VersionBumper arithmetic
// ─────────────────────────────────────────────────────────────────────────────

public static class AppVersionBumper
{
    public static AppSemanticVersion Bump(AppSemanticVersion v, AppBumpType t) => t switch
    {
        AppBumpType.Major => new AppSemanticVersion(v.Major + 1, 0, 0),
        AppBumpType.Minor => new AppSemanticVersion(v.Major, v.Minor + 1, 0),
        AppBumpType.Patch => new AppSemanticVersion(v.Major, v.Minor, v.Patch + 1),
        AppBumpType.None  => v,
        _                 => throw new ArgumentOutOfRangeException(nameof(t), t, null)
    };
}

// ─────────────────────────────────────────────────────────────────────────────
// ChangelogGenerator
// ─────────────────────────────────────────────────────────────────────────────

public static class AppChangelogGenerator
{
    public static string Generate(
        AppSemanticVersion version,
        IEnumerable<AppCommit> commits,
        DateTime? date = null)
    {
        var releaseDate = date ?? DateTime.UtcNow;
        var list = commits.ToList();

        var sb = new System.Text.StringBuilder();
        sb.AppendLine($"## [{version}] - {releaseDate:yyyy-MM-dd}");

        var breaking = list.Where(c => c.IsBreaking).ToList();
        var feats    = list.Where(c => !c.IsBreaking && c.Type == "feat").ToList();
        var fixes    = list.Where(c => !c.IsBreaking && c.Type == "fix").ToList();
        var other    = list.Where(c => !c.IsBreaking
                                     && c.Type != "feat"
                                     && c.Type != "fix"
                                     && c.Type != "other").ToList();

        AppendSection(sb, "BREAKING CHANGES", breaking);
        AppendSection(sb, "Features",         feats);
        AppendSection(sb, "Bug Fixes",        fixes);
        AppendSection(sb, "Other Changes",    other);

        return sb.ToString().TrimEnd() + Environment.NewLine;
    }

    private static void AppendSection(System.Text.StringBuilder sb, string title, List<AppCommit> commits)
    {
        if (commits.Count == 0) return;
        sb.AppendLine();
        sb.AppendLine($"### {title}");
        sb.AppendLine();
        foreach (var c in commits)
            sb.AppendLine($"- {(c.Scope != null ? $"**{c.Scope}**: " : "")}{c.Description}{(c.IsBreaking ? " ⚠️ BREAKING" : "")}");
    }
}

File: results/2026-05-06_173435/tasks/11-semantic-version-bumper/default-sonnet46-1m-medium/generated-code/bump_version.py
"""
Semantic Version Bumper

Reads a VERSION file (or package.json), analyzes conventional commits to determine
the appropriate semver bump (major/minor/patch), updates the file, and generates
a changelog entry.

Conventional commit rules:
  BREAKING CHANGE or `type!:` in any commit -> major
  feat:                                      -> minor
  fix: (or any other type)                   -> patch
"""
import sys
import re
import os
import json
from pathlib import Path


def parse_version(version_str: str) -> tuple[int, int, int]:
    """Parse 'MAJOR.MINOR.PATCH' string into a tuple of ints."""
    match = re.match(r"^(\d+)\.(\d+)\.(\d+)\s*$", version_str.strip())
    if not match:
        raise ValueError(f"Invalid version format: {version_str!r}")
    return int(match.group(1)), int(match.group(2)), int(match.group(3))


def determine_bump_type(commits: list[str]) -> str:
    """
    Scan commit messages and return the highest required bump level.
    Short-circuits on 'major'; otherwise accumulates highest seen level.
    """
    bump = "patch"
    for commit in commits:
        line = commit.strip()
        # Breaking change: explicit keyword anywhere in message
        if "BREAKING CHANGE" in line:
            return "major"
        # Breaking change: type followed by ! before colon (e.g. feat!, fix!, feat(scope)!)
        if re.match(r"^[a-z]+(\([^)]+\))?!:", line):
            return "major"
        # Feature bump
        if re.match(r"^feat(\([^)]+\))?:", line):
            bump = "minor"
    return bump


def bump_version(version: tuple[int, int, int], bump_type: str) -> tuple[int, int, int]:
    """Return new version tuple after applying bump_type."""
    major, minor, patch = version
    if bump_type == "major":
        return (major + 1, 0, 0)
    if bump_type == "minor":
        return (major, minor + 1, 0)
    return (major, minor, patch + 1)


def format_version(version: tuple[int, int, int]) -> str:
    return ".".join(str(v) for v in version)


def generate_changelog_entry(version: str, commits: list[str]) -> str:
    """Build a markdown changelog section for the given version."""
    breaking: list[str] = []
    features: list[str] = []
    fixes: list[str] = []
    other: list[str] = []

    for commit in commits:
        line = commit.strip()
        if not line:
            continue
        if "BREAKING CHANGE" in line or re.match(r"^[a-z]+(\([^)]+\))?!:", line):
            breaking.append(f"- {line}")
        elif re.match(r"^feat(\([^)]+\))?:", line):
            features.append(f"- {line}")
        elif re.match(r"^fix(\([^)]+\))?:", line):
            fixes.append(f"- {line}")
        else:
            other.append(f"- {line}")

    sections: list[str] = [f"## {version}\n"]
    if breaking:
        sections.append("### Breaking Changes\n")
        sections.extend(breaking)
        sections.append("")
    if features:
        sections.append("### Features\n")
        sections.extend(features)
        sections.append("")
    if fixes:
        sections.append("### Bug Fixes\n")
        sections.extend(fixes)
        sections.append("")
    if other:
        sections.append("### Other\n")
        sections.extend(other)
        sections.append("")

    return "\n".join(sections) + "\n"


def read_version(version_file: str) -> str:
    """Read version string from VERSION file or package.json."""
    p = Path(version_file)
    if p.name == "package.json":
        data = json.loads(p.read_text())
        return data["version"]
    return p.read_text().strip()


def write_version(new_version: str, version_file: str) -> None:
    """Write new version to VERSION file or package.json."""
    p = Path(version_file)
    if p.name == "package.json":
        data = json.loads(p.read_text())
        data["version"] = new_version
        p.write_text(json.dumps(data, indent=2) + "\n")
    else:
        p.write_text(new_version + "\n")


def read_commits(commits_file: str) -> list[str]:
    """Read non-blank commit messages from a text file (one per line)."""
    p = Path(commits_file)
    if not p.exists():
        return []
    return [line for line in p.read_text().splitlines() if line.strip()]


def main() -> None:
    version_file = os.environ.get("VERSION_FILE", "VERSION")
    commits_file = os.environ.get("COMMITS_FILE", "commits.txt")

    if not Path(version_file).exists():
        print(f"Error: version file not found: {version_file}", file=sys.stderr)
        sys.exit(1)

    current = read_version(version_file)
    try:
        version_tuple = parse_version(current)
    except ValueError as exc:
        print(f"Error: {exc}", file=sys.stderr)
        sys.exit(1)

    commits = read_commits(commits_file)
    if not commits:
        print(f"Warning: no commits found in {commits_file}", file=sys.stderr)

    bump_type = determine_bump_type(commits)
    new_tuple = bump_version(version_tuple, bump_type)
    new_version = format_version(new_tuple)

    write_version(new_version, version_file)

    # Prepend new entry to CHANGELOG.md
    changelog_path = Path("CHANGELOG.md")
    existing = changelog_path.read_text() if changelog_path.exists() else ""
    entry = generate_changelog_entry(new_version, commits)
    changelog_path.write_text(entry + existing)

    print(f"Bumped {current} -> {new_version} ({bump_type})")
    print(f"NEW_VERSION={new_version}")


if __name__ == "__main__":
    main()
